### PR TITLE
Visualize Empire Spawn Pressure

### PR DESCRIPTION
**Pull request by jschavey**
Adds an indicator to global visualization for empire-wide spawn pressure.

### CODE REVIEW

File: visuals.js [hunk 1]
```diff
@@ -58,7 +58,34 @@ Visuals.extend = function() {
...
+                let count = 0;
+                for(let roomName in Game.rooms) {
+                    const room = Game.rooms[roomName];
+                    count += _.size(room.spawnQueueHigh);
+                    count += _.size(room.spawnQueueMedium);
+                    count += _.size(room.spawnQueueLow);
+                }
+                return count;
+            }
+            vis.rect(x, y + 0.75, sectionWidth, 1, BAR_STYLE);
```
Comment (Spedwards) [inline visuals.js]: `y - 0.75`

File: visuals.js [hunk 2]
```diff
@@ -58,7 +58,34 @@ Visuals.extend = function() {
...
+                let count = 0;
+                for (let spawnName in Game.spawns) {
+                    if (Game.spawns[spawnName].spawning !== null) count++;
+                }
+                return count;
+            }
+            let numToSpawn = function () {
+                let count = 0;
+                for(let roomName in Game.rooms) {
+                    const room = Game.rooms[roomName];
```
Comment (Spedwards) [inline visuals.js]: `_.forEach(Game.rooms, room => {`

File: visuals.js [hunk 3]
```diff
@@ -58,7 +58,34 @@ Visuals.extend = function() {
                 opacity: BAR_STYLE.opacity
             });
             vis.text(`GCL: ${Game.gcl.level} (${(GCL_PERCENTAGE * 100).toFixed(2)}%)`, x + sectionWidth / 2, y);
-            
+            //  Spawn Pressure
+            let spawnCount = _.size(Game.spawns);
+            let numSpawnsSpawning = function () {
+                let count = 0;
+                for (let spawnName in Game.spawns) {
+                    if (Game.spawns[spawnName].spawning !== null) count++;
```
Comment (Spedwards) [inline visuals.js]: `count += _(Game.spawns).reject('spawning').size();`

File: visuals.js [hunk 4]
```diff
@@ -58,7 +58,34 @@ Visuals.extend = function() {
...
+                    count += _.size(room.spawnQueueHigh);
+                    count += _.size(room.spawnQueueMedium);
+                    count += _.size(room.spawnQueueLow);
+                }
+                return count;
+            }
+            vis.rect(x, y + 0.75, sectionWidth, 1, BAR_STYLE);
+            let pressure = (numSpawnsSpawning() + numToSpawn()) / spawnCount;
+                pressure = Math.min(1, pressure);
+            vis.rect(x, y + 0.75, pressure * sectionWidth, 1, {
```
Comment (Spedwards) [inline visuals.js]: `y - 0.75`

File: visuals.js [hunk 5]
```diff
@@ -58,7 +58,34 @@ Visuals.extend = function() {
...
+                fill: getColourByPercentage(pressure),
+                opacity: BAR_STYLE.opacity
+            });
+            vis.text(`Spawn Pressure: ${(pressure * 100).toFixed(2)}%`, x + sectionWidth / 2, y + 1.5);
```
Comment (Spedwards) [inline visuals.js]: Don't add to `y`?

File: visuals.js [hunk 6]
```diff
@@ -58,7 +58,34 @@ Visuals.extend = function() {
...
+                }
+                return count;
+            }
+            let numToSpawn = function () {
+                let count = 0;
+                for(let roomName in Game.rooms) {
+                    const room = Game.rooms[roomName];
+                    count += _.size(room.spawnQueueHigh);
+                    count += _.size(room.spawnQueueMedium);
+                    count += _.size(room.spawnQueueLow);
```
Comment (Spedwards) [inline visuals.js]: `count += _(Game.rooms).map(r => r.spawnQueueHigh.concat(r.spawnQueueMedium, r.spawnQueueLow)).size()` is another option.

**Comment by jschavey** [comment]
requested changes made

File: visuals.js [hunk 7]
```diff
@@ -83,6 +83,33 @@ Visuals.extend = function() {
...
+            }
+            const numToSpawn = function () {
+                let count = 0;
+                _.forEach (Game.rooms, room => {
+                    count += _.size(room.spawnQueueHigh);
+                    count += _.size(room.spawnQueueMedium);
+                    count += _.size(room.spawnQueueLow);
+                })
+                return count;
+            }
```
Comment (Spedwards) [inline visuals.js]: Replace both functions with the following:

```
let count = _(Game.spawns).reject('spawning').size();
count += _(Game.rooms).map(r => r.spawnQueueHigh.concat(r.spawnQueueMedium, r.spawnQueueLow)).size();
```

File: visuals.js [hunk 8]
```diff
@@ -83,6 +83,33 @@ Visuals.extend = function() {
...
+            vis.rect(x, y - 0.75, sectionWidth, 1, BAR_STYLE);
+            let pressure = (numSpawnsSpawning() + numToSpawn()) / spawnCount;
+                pressure = Math.min(1, pressure);
```
Comment (Spedwards) [inline visuals.js]: Don't do this. `const PRESSURE = Math.min(1, count / spawnCount);`

**Comment by jschavey** [comment]
My initial labeling is ambiguous.  I meant "Spawn Pressure" in that if your empire is >100% pressured it is spawn locked.  As I use this, however, I have come to think of spawn pressure semantically as instantaneous spawning capacity.  Would need to alter the algorithm to support this.  Thoughts?

**Comment by karlthepagan** [comment]
Each creep has a spawn cost of 3* part count / 1500 (or 500 for reserve & claim)

You can either measure it that way (total cost over time) but that means totaling creeps + queued 

The other way is the current spawn queue depth in ticks.

**Comment by chobobobo** [comment]
@jschavey not sure myself where this code fits so I can't resolve the conflict, but once it's ready I will put it in canary to try it out

**Comment by jschavey** [comment]
updated with dev, uses new helper methods drawBar and drawPie

File: visuals.js [hunk 9]
```diff
@@ -313,17 +313,7 @@ const Visuals = class {
...
-                        const room = Game.rooms[roomName];
-                        count += _.size(room.spawnQueueHigh);
-                        count += _.size(room.spawnQueueMedium);
-                        count += _.size(room.spawnQueueLow);
-                    }
-                    return count;
-                }
-                count += globalQueueCount();
+                count += _(Game.rooms).map(r => r.spawnQueueHigh.concat(r.spawnQueueMedium, r.spawnQueueLow)).flatten().size();
                 const SCU_PERCENTAGE = count / spawnCount;
```
Comment (jschavey) [inline visuals.js]: The inclusion of .flatten() appears to be what was needed to bring this one-liner into consistency with the replaced function.

**Comment by jschavey** [comment]
I've observed Spawn Capacity Utilization get as high as 360%.  I personally prefer the stylization of two significant figures with a single precision digit when considering this range.  I can, of course, still make this change if you feel its necessary.

File: visuals.js [hunk 10]
```diff
@@ -320,8 +334,15 @@ const Visuals = class {
             // BUCKET
             this.drawPie(vis, Game.cpu.bucket, 10000, 'Bucket', getColourByPercentage(Math.min(1, Game.cpu.bucket / 10000), true), {x, y: y++});
             
+            //  SPAWN CAPACITY UTILIZATION (SCU)
+            const spawnCount = _.size(Game.spawns);
+            let count = _(Game.spawns).filter('spawning').size();
+            count += _(Game.rooms).map(r => r.spawnQueueHigh.concat(r.spawnQueueMedium, r.spawnQueueLow)).flatten().size();
+            const SCU_PERCENTAGE = count / spawnCount;
+            this.drawPie(vis, SCU_PERCENTAGE, 1, 'SCU', getColourByPercentage(SCU_PERCENTAGE), {x, y: y++});
```
Comment (Spedwards) [inline visuals.js]: You need to run `Math.floor` or `Math.round` on the second argument at this time. Otherwise you get wrong results displayed.

**Pull request merged**